### PR TITLE
document the duck array integration status

### PR DESCRIPTION
**Pull request by keewis**
With the initial (experimental) release of `pint-xarray` getting closer (see xarray-contrib/pint-xarray#25), I figured it might make sense to document the parts of the API of `xarray` which right now won't work with duck arrays, and to start maintaining a list of extensions like `pint-xarray`.

The remaining issues are taken from #3594, but I'm sure I missed something (e.g. plotting).

 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: doc/duckarrays.rst [hunk 1]
```diff
@@ -33,4 +33,12 @@ will still cast to ``numpy`` arrays:
...
\ No newline at end of file
+    :py:func:`numpy.vectorize`)
+
+
+Extensions using duck arrays
+----------------------------
+Here's a list of libraries extending ``xarray`` to make working with wrapped duck arrays
+easier:
+
+- `pint-xarray <https://github.com/xarray-contrib/pint-xarray>`_
```
Comment (dcherian) [inline doc/duckarrays.rst:44]: @jacobtomlinson do you want to list `cupy_xarray` here or is it too early to do that?

**Comment by keewis** [comment]
I'm not sure "plotting" should be included in that list since converting to numpy is intentional.

Other than that, this should be ready for review.

File: doc/duckarrays.rst [hunk 2]
```diff
@@ -1,7 +1,7 @@
 .. currentmodule:: xarray
 
-Using duck arrays with xarray
-=============================
+Working with duck arrays
```
Comment (dcherian) [inline doc/duckarrays.rst:3]: I think "duck array" is a little technical (hah). Shall we call this "Working with numpy-like arrays"?

Comment (keewis) [reply doc/duckarrays.rst:3]: I agree, it's probably better to use a less technical heading and then mention the technical term in the introduction?

Comment (dcherian) [reply doc/duckarrays.rst:3]: :+1:

File: doc/duckarrays.rst [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+.. warning::
+
+   This is a experimental feature, please report any bugs you might find.
+
+Numpy-like arrays (:term:`duck array`) extend the :py:class:`numpy.ndarray` with
+additional features, like propagating physical units or a different layout in memory.
+
+:py:class:`DataArray` and :py:class:`Dataset` objects can wrap these duck arrays, as
+long as they satisfy certain conditions (see :ref:`internals.duck_arrays`).
```
Comment (keewis) [inline doc/duckarrays.rst:14]: would it make sense to point to the explicitly tested duck arrays here (`pint`, `sparse`)? We could also add a user-maintained list of duck array libraries, just like the current "related projects" list.

I did think about adding usage examples, but maybe it's better to leave that to the extension packages?

File: doc/terminology.rst [hunk 4]
```diff
@@ -104,3 +104,11 @@ complete examples, please consult the relevant documentation.*
         one, it has 0 dimensions. That means that, e.g., :py:class:`int`,
         :py:class:`float`, and :py:class:`str` objects are "scalar" while
         :py:class:`list` or :py:class:`tuple` are not.
+
+    duck array
+        `Duck arrays <duck array>`_ are array implementations that behave
+        like numpy arrays. They have to define the ``shape``, ``dtype`` and
+        ``ndim`` properties. For integration with ``xarray``, the ``__array__``,
+        ``__array_ufunc__`` and ``__array_function__`` protocols are also required.
```
[the remark this answers is not in the document]
Comment (dcherian) [reply doc/terminology.rst:112]: :+1: from me...

File: doc/duckarrays.rst [hunk 5]
```diff
@@ -0,0 +1,54 @@
+.. currentmodule:: xarray
+
+Working with numpy-like arrays
```
Comment (max-sixty) [inline doc/duckarrays.rst:3]: Is it worth referencing that xarray handles dask arrays natively?

**Review by max-sixty — APPROVED** [review]
Great idea, thanks @keewis

File: doc/duckarrays.rst (hunk 5 above)
Comment (keewis) [reply doc/duckarrays.rst:3]: I added a note. Does that help?

**Comment by keewis** [comment]
I just noticed that we have to wrap `chunk` calls with a `dequantify` / `quantify` cycle if we want to have it work correctly, but maybe that's something that belongs into the `pint-xarray` docs?

File: doc/duckarrays.rst (hunk 1 above)
Comment (jacobtomlinson) [reply doc/duckarrays.rst:59]: I would say it's too early for that. But good to know where to put it.

**Comment by keewis** [comment]
if there are no further comments, should we merge this? I'm almost certain the list of missing features is not complete, but I would open other PRs once new items are found.

File: doc/terminology.rst [hunk 6]
```diff
@@ -104,3 +104,11 @@ complete examples, please consult the relevant documentation.*
         one, it has 0 dimensions. That means that, e.g., :py:class:`int`,
         :py:class:`float`, and :py:class:`str` objects are "scalar" while
         :py:class:`list` or :py:class:`tuple` are not.
+
+    duck array
+        `Duck arrays <duck array>`_ are array implementations that behave
```
Comment (jhamman) [inline doc/terminology.rst:109]: I get a 404 from this link when I try to view the new page on the readthedocs PR build. Are we sure this is right?

Comment (jhamman) [inline doc/terminology.rst:109]: ?
```suggestion
        `Duck arrays <duckarrays>`_ are array implementations that behave
```

Comment (keewis) [reply doc/terminology.rst]: thanks for catching this, it seems we can't use the `text <label>` syntax in combination with labels. Should be fixed now.

File: doc/duckarrays.rst (hunk 3 above)
Comment (dcherian) [reply doc/duckarrays.rst:15]: I think a list would be nice.

**Review by dcherian — APPROVED** [review]
Thanks!

**Pull request merged**